server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2648,4 +2645,148 @@ public void afterRefresh(boolean didRefresh) throws IOException {
+     */
+    static final class EngineHolder implements Closeable {
+        private volatile Engine activeEngine; // allow to read without synchronization
+        private Engine pendingEngine;
+        // avoid losing acknowledged writes, only activate the pending engine when its local_checkpoint at least this guard.
+        private long minRequiredCheckpoint = NO_OPS_PERFORMED;
+
+        synchronized void activateNewEngine(Engine newEngine) {
+            assert activeEngine == null : "engine is activated already";
+            assert pendingEngine == null : "engine is being reset";
+            this.activeEngine = newEngine;
+        }
+
+        Engine getActiveEngineOrNull() {
+            return activeEngine;
+        }
+
+        synchronized void makeActiveEngineSearchOnly() throws IOException {
+            assert activeEngine != null : "engine is not activated yet";
+            if ((this.activeEngine instanceof SearchOnlyEngine) == false) {
+                final Engine current = this.activeEngine;
+                this.activeEngine = new SearchOnlyEngine(this.activeEngine);
+                IOUtils.close(current);
+            }
+        }
+
+        synchronized Engine getEngineForResync() {
+            final Engine target = pendingEngine != null ? pendingEngine : activeEngine;
+            if (target == null) {
+                throw new AlreadyClosedException("engine is closed");
+            }
+            return target;
+        }
+
+        synchronized boolean hasPendingEngine() {
+            return pendingEngine != null;
+        }
+
+        synchronized void setPendingEngine(Engine pendingEngine, long minRequiredCheckpoint) {
+            assert this.activeEngine != null : "engine is not activated";
+            assert this.pendingEngine == null : "engine is being reset already";
+            this.pendingEngine = pendingEngine;
+            this.minRequiredCheckpoint = minRequiredCheckpoint;
+        }
+
+        synchronized void adjustMinRequiredCheckpoint(long newMinRequiredCheckpoint) {
+            this.minRequiredCheckpoint = Math.min(minRequiredCheckpoint, newMinRequiredCheckpoint);
+        }
+
+        synchronized void closePendingEngine() throws IOException {
+            final Engine closing = this.pendingEngine;
+            this.pendingEngine = null;
+            IOUtils.close(closing);
+        }
+
+        synchronized void maybeActivatePendingEngine() throws IOException {
+            if (pendingEngine != null && pendingEngine.getLocalCheckpoint() >= minRequiredCheckpoint) {
+                // make sure that all acknowledged writes are visible
+                pendingEngine.refresh("rollback");
+                final Engine closing = this.activeEngine;
+                this.activeEngine = this.pendingEngine;
+                this.pendingEngine = null;
+                IOUtils.close(closing);
+            }
+        }
+
+        @Override
+        public void close() throws IOException {
```

Comment:
make this synchronized too. it's safer since you modify both references

Reply:
good catch!